volMgr/volMgr.go

```diff
@@ -87,26 +91,46 @@ func (m *vmgr) CreateVol(id, rootfs, mountpoint string, uid, gid uint32, shiftUi
 	var err error
 
 	volPath := filepath.Join(m.hostDir, id)
-	mountPath := filepath.Join(rootfs, mountpoint)
-
 	if _, err = os.Stat(volPath); err == nil {
 		return nil, fmt.Errorf("volume dir for container %v already exists", id)
 	}
 
+	mountPath := filepath.Join(rootfs, mountpoint)
+
+	rootfsOnOvfs, rootfsOvfsUpper, err := isRootfsOnOverlayfs(rootfs)
+	if err != nil {
+		return nil, err
+	}
+
+	// When the container stops and we need to copy the volume contents back to
+	// the container's rootfs. We call this "sync-out", and syncOutPath is the
+	// path were we want to copy to.
+	syncOutPath := mountPath
+
+	// If the container rootfs is on overlayfs, the syncOutPath can't be the
```

Comment:
Good questions. 

> My understanding is that once an ovfs resource is mounted, all user I/O is expected to be targeted at the merge mountpoint, and that only kernel is the one expected to be writing into upper/lower layers. Just wondering if this is a fully support operation that can't / won't give us problems down the road.

That's correct; but we have no other choice as the merged dir is ID-mapped mounted only within the container's mount namespace, so sysbox-mgr can't copy files to/from it. Our only choice is to copy to the "upper" dir, and works fine because when we copy the overlayfs mount on the merged this is not present yet (i.e., overlayfs is not using the upper dir yet). 

As an aside, recall that sysbox-mgr copies the files in `/var/lib/sysbox/docker/<container-uuid>` (and similar) to/from the container's rootfs when the container pauses or stops, so that `docker commit` can capture inner docker images (and this also makes image preloading with `docker build` possible).

> In regards to the above, you say "sysbox-runc may". Is sysbox-runc actually doing this all the time when ovfs-over-idmapped is supported, or only in certain occasions? And if so, when? I would expect a binary answer here ("always" or "never"), but looks like it depends based on your comment.

I said "sysbox-runc may have remounted" because when the kernel does not support overlayfs on ID-mapped mounts, then sysbox won't remount the rootfs (e.g., kernels < 5.19, or when docker is in user-ns remap mode, or when using CRI-O + sysbox). 

In all of those scenarios the copying of data from `/var/lib/sysbox/docker/<container-uuid>` to the container's rootfs must still work. In those other scenarios we could copy the data to the rootfs `merged` dir (as we were doing before this PR), but that would complicate the code even further as it would mean some scenarios copy to the rootfs merged dir, others copy to the rootfs upper dir. The latter always works for all scenarios, so I chose that one.

Reply:
Makes sense.